projects/igniteui-angular/src/lib/time-picker/time-picker.component.ts

```diff
@@ -2124,71 +1742,187 @@ export class IgxTimePickerComponent implements
+        const minutes = value.getMinutes();
+        const formattedMinutes = minutes < 10 && format.indexOf('mm') !== -1 ? '0' + minutes : `${minutes}`;
+        return formattedMinutes;
+    }
+
+    private parseSeconds(value: Date, format: string): string {
+        const seconds = value.getSeconds();
+        const formattedSeconds = seconds < 10 && format.indexOf('ss') !== -1 ? '0' + seconds : `${seconds}`;
+        return formattedSeconds;
+    }
+
+    private parseAMPM(value: Date, format: string): string {
+        let hour = value.getHours();
+        const ampm = hour >= 12 ? 'PM' : 'AM';
+        return ampm;
+    }
+
+    private handleContainerClosed(): void {
+        const newValue = this.dateTimeEditor.value;
+        const minDate: Date = this.parseToDate(this.minValue);
+        const maxDate: Date = this.parseToDate(this.maxValue);
+
+        if (this.value.getTime() === newValue.getTime()) {
             return;
         }
 
-        if (this._isValueValid(newVal)) {
-            if (!this.value || oldValue.getTime() !== newVal.getTime()) {
-                this.value = newVal;
-            }
+        if (this.valueInRange(newValue, minDate, maxDate)) {
+            this.value = newValue;
         } else {
-            this.displayValue = this.inputFormat.transform(this._formatTime(oldValue, this.format));
-
-            const args: IgxTimePickerValidationFailedEventArgs = {
-                timePicker: this,
-                currentValue: newVal,
-                setThroughUI: true
-            };
-            this.onValidationFailed.emit(args);
+            this.emitValidationFailedEvent(newValue, true);
         }
     }
 
-    private cursorOnHours(cursor: number, showHours: boolean): boolean {
-        return showHours && this._hoursPos.has(cursor);
+    private shouldCancelSelecting(): boolean {
+        const args: IBaseCancelableBrowserEventArgs = { owner: this, cancel: false };
+        this.selecting.emit(args);
+        return args.cancel;
     }
 
-    private cursorOnMinutes(cursor: number, showHours: boolean, showMinutes: boolean): boolean {
-        return showMinutes &&
-            (showHours && this._minutesPos.has(cursor)) ||
-            (!showHours && this._minutesPos.has(cursor));
+    private emitValueChangedEvent(oldValue: Date, newValue: Date) {
+        const args: IgxTimePickerValueChangedEventArgs = {
+            oldValue: oldValue,
+            newValue: newValue
+        };
+        this.valueChanged.emit(args);
     }
 
-    private cursorOnSeconds(cursor: number, showHours: boolean, showMinutes: boolean, showSeconds: boolean): boolean {
-        return showSeconds &&
-            (showHours && showMinutes && this._secondsPos.has(cursor)) ||
-            ((!showHours || !showMinutes) && this._secondsPos.has(cursor)) ||
-            (!showHours && !showMinutes && this._secondsPos.has(cursor));
+    private emitValidationFailedEvent(currentValue: Date, setThroughUI = false) {
```

Comment:
```suggestion
    private emitValidationFailedEvent(currentValue: Date | string, setThroughUI = false) {
```

Reply:
Also , we may want to rethink `IgxTimePickerValidationFailedEventArgs`. We could define common event args emitted by the three pickers.